adding anyParams directive as used to exist in spray.io, issue #526

nairbv opened the pull request:
Attempting to add an anyParams directive as exists in spray.io for issue #526 
https://github.com/akka/akka-http/issues/526

This is just merging the regular parameter and form field directives, with some liberal copy/pasting of some magnet patterns I'm not an expert on.

I didn't include all of the ParameterDirectiveSpec tests because they seemed redundant.  A few of the FormFieldDirectiveSpec tests didn't pass when run against anyParams, and so were excluded, because a query parameter can't accept certain types that form fields can.

lightbend-cla-validator commented:
Hi @nairbv,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

ktoso commented:
OK TO TEST

jrudolph commented:
Cool, thanks for tackling this. In https://github.com/akka/akka-http/pull/1188 I attempted a cleanup of the implicits around form fields. It would be nice if we could do this simplification for AnyParams directly. Could you try if you can pull that off? The idea is to remove all FSU and FSFFU implicit parameters and use the statically known unmarshallers directly.

nairbv commented:
ah, I added the CLA signature, but haven't looked at the java version of these directives so some validation check is still failing.

@jrudolph I dropped the sfu parameter, and it was simple enough to remove the fsu:FSU[String], but stringFieldUnmarshaller is only available in the version of StrictForm in your branch right?  I assume it will conflict if I make the same change in my branch.  It looks like you don't plan to remove the XXX[T] implicits right, only those parameterized on String?

nairbv commented:
Where can I find info on running the PR validation locally?  Building with 'sbt test' succeeds on my desktop after these changes.  Not sure what it's saying is binary incompatible...

nairbv commented:
@ktoso @jrudolph anything else that should be changed?

jlprat commented:
Hi @nairbv, thanks for the PR!
Summertime is a little slower than usual due to people being on vacation.
One think I've seen is that documentation for this new set of directives is missing. Could you also provide it? Feel free to ask if you have any questions :smile:
I'll try to take a deeper look at your PR today.

raboof commented:
Seems neat! @nairbv would you be willing to also add this directive to the documentation?

raboof commented:
Hmm, 'Error: Unable to access jarfile /usr/share/sbt-launcher-packaging/bin/sbt-launch.jar'?

raboof commented:
PLS BUILD

raboof commented:
The build now fails due to:
```
[info] akka-http: found 5 potential binary incompatibilities while checking against com.typesafe.akka:akka-http_2.11:10.0.10  (filtered 3)
[error]  * abstract method anyParamMultiMap()akka.http.scaladsl.server.Directive in trait akka.http.scaladsl.server.directives.AnyParamDirectives is inherited by class Directives in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.http.scaladsl.server.directives.AnyParamDirectives.anyParamMultiMap")
[error]  * abstract method anyParamMap()akka.http.scaladsl.server.Directive in trait akka.http.scaladsl.server.directives.AnyParamDirectives is inherited by class Directives in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.http.scaladsl.server.directives.AnyParamDirectives.anyParamMap")
[error]  * abstract method anyParams(akka.http.scaladsl.server.directives.AnyParamDirectives#AnyParamMagnet)java.lang.Object in trait akka.http.scaladsl.server.directives.AnyParamDirectives is inherited by class Directives in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.http.scaladsl.server.directives.AnyParamDirectives.anyParams")
[error]  * abstract method anyParamSeq()akka.http.scaladsl.server.Directive in trait akka.http.scaladsl.server.directives.AnyParamDirectives is inherited by class Directives in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.http.scaladsl.server.directives.AnyParamDirectives.anyParamSeq")
[error]  * abstract method anyParam(akka.http.scaladsl.server.directives.AnyParamDirectives#AnyParamMagnet)java.lang.Object in trait akka.http.scaladsl.server.directives.AnyParamDirectives is inherited by class Directives in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.http.scaladsl.server.directives.AnyParamDirectives.anyParam")
```

@nairbv you can run the pull request validation locally as documented in https://github.com/akka/akka/blob/master/CONTRIBUTING.md#pull-request-validation-workflow-details

nairbv commented:
Strange, I think tests passed last time I tried. Will take a look when I get some time but not sure when that will be.

jrudolph commented:
PLS BUILD

jrudolph commented on akka-http/src/main/scala/akka/http/javadsl/server/Directives.scala:
```diff
@@ -8,11 +8,13 @@ import java.util.function.{ BiFunction, Function, Supplier }
 import scala.annotation.varargs
 
-abstract class AllDirectives extends FramedEntityStreamingDirectives
+abstract class AllDirectives extends AnyParamDirectives
```
I think that one is wrong. You need to add the extension to BasicDirectives, instead and keep, WebSocketDirectives here (though, currently it is weird because someone broke up the alphabetic order here).

jrudolph commented:
Sorry for keeping that around for so long, @nairbv. By now the FormFieldDirectives changes from #1188 have been merged. It would be good if you good adapt this PR to use the same set of simplified magnets as there.

Apart from that I think the PR looks quite good. It's a lot of added code, though, so we should mark the new API with `@ApiMayChange` if we have missed some unforeseen consequences.

jrudolph commented:
Btw, if you don't want to continue on this, @nairbv, please let us know so we can stash this PR for now until someone else wants to pick it up again.

jrudolph commented:
This still needs a bit of work, so closing for now.
